## Gather

In [216]:
import pandas as pd
import numpy as np

In [217]:
patients = pd.read_csv('C:/Users/ENTER/Desktop/n-p/1/home/patients.csv')
treatments = pd.read_csv('C:/Users/ENTER/Desktop/n-p/1/home/treatments.csv')
adverse_reactions = pd.read_csv('C:/Users/ENTER/Desktop/n-p/1/home/adverse_reactions.csv')

## Assess

#### 1- patients data

In [218]:
patients.sample(3)

patient_id assigned_sex given_name  surname             address  \
25           26         male     Gregor     Bole   922 Chapmans Lane   
495         496         male     Hajime  Tsukada   4111 Thunder Road   
26           27       female       Ella     Lund  1207 Garfield Road   

            city state  zip_code        country  \
25   Albuquerque    NM   87109.0  United States   
495    San Mateo    CA   94403.0  United States   
26        Peoria    IL   61602.0  United States   

                                  contact   birthdate  weight  height   bmi  
25       GregorBole@gustr.com505-828-4955   6/19/1922   180.8      67  28.3  
495  650-570-4896HajimeTsukada@dayrep.com    9/5/1972   168.1      66  27.1  
26       309-671-8852EllaLund@armyspy.com  12/19/1933   144.8      61  27.4

In [219]:
patients.columns

Index(['patient_id', 'assigned_sex', 'given_name', 'surname', 'address',
       'city', 'state', 'zip_code', 'country', 'contact', 'birthdate',
       'weight', 'height', 'bmi'],
      dtype='object')

In [220]:
patients.shape

(503, 14)

In [221]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 37.4+ KB


In [222]:
patients.duplicated().sum()

0

In [223]:
patients.dtypes

patient_id        int64
assigned_sex     object
given_name       object
surname          object
address          object
city             object
state            object
zip_code        float64
country          object
contact          object
birthdate        object
weight          float64
height            int64
bmi             float64
dtype: object

In [224]:
var=list(patients.columns)
for item in var:
    null_value_incol = patients[item].isnull().sum()
    print(item+":"+str(null_value_incol))

patient_id:0
assigned_sex:0
given_name:0
surname:0
address:12
city:12
state:12
zip_code:12
country:12
contact:12
birthdate:0
weight:0
height:0
bmi:0


In [225]:
#for manually deep look for sample
patients.sample(8)

patient_id assigned_sex  given_name        surname               address  \
39           40       female    Ganimete        Ščančar  4105 Ferguson Street   
222         223       female      Malina        Balagić     4019 Cerullo Road   
187         188         male   Władysław      Wieczorek  2338 Virginia Street   
56           57       female      Alissa         Montez     1435 Limer Street   
42           43       female      Mizuki          Iwata      3434 Holt Street   
81           82         male       André         Nordin  836 Dovetail Estates   
460         461         male        Arne      Jørgensen     1033 Wayside Lane   
135         136         male  Willem-Jan  van der Lubbe   1717 Vineyard Drive   

                city state  zip_code        country  \
39           Walpole    MA    2081.0  United States   
222       Louisville    KY   40244.0  United States   
187          Chicago    IL   60605.0  United States   
56          Eatonton    GA   31024.0  United States   
42   West Palm Beach    FL   33409.0  United States   
81              Gate    OK   73844.0  United States   
460          Oakland    CA   94612.0  United States   
135        Cleveland    OH   44115.0  United States   

                                           contact   birthdate  weight  \
39            508-454-2027GanimeteScancar@cuvox.de  10/25/1947   191.6   
222           502-367-6167MalinaBalagic@einrot.com   2/14/1962   224.8   
187     773-607-2647WladyslawWieczorek@teleworm.us   1/15/1994   133.1   
56   AlissaMontezFranco@gustr.com+1 (706) 497-0891    1/8/1951   133.8   
42               MizukiIwata@gustr.com561-357-0702   6/30/1927   199.1   
81                580-934-1141AndreNordin@cuvox.de  11/12/1940   209.4   
460      ArneJrgensen@fleckens.hu+1 (510) 238-1151   3/14/1998   163.5   
135    Willem-JanvanderLubbe@gustr.com440-385-5011    7/9/1941   152.9   

     height   bmi  
39       67  30.0  
222      68  34.2  
187      67  20.8  
56       61  25.3  
42       61  37.6  
81       69  30.9  
460      65  27.2  
135      69  22.6

In [226]:
var=list(patients.columns)
for item in var:
    duplicate=patients[item].duplicated().sum()
    print(item +":"+str(duplicate))

patient_id:0
assigned_sex:501
given_name:33
surname:37
address:19
city:153
state:448
zip_code:72
country:501
contact:19
birthdate:10
weight:156
height:481
bmi:320


In [227]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [228]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [229]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [230]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [231]:
patients.surname.value_counts()

Doe            6
Taylor         3
Jakobsen       3
Hueber         2
Lund           2
              ..
Rukavina       1
Konovšek       1
Kozłowska      1
Villalpando    1
Vukelić        1
Name: surname, Length: 466, dtype: int64

In [232]:
patients.address.value_counts()

123 Main Street         6
2476 Fulton Street      2
2778 North Avenue       2
648 Old Dear Lane       2
4984 Hampton Meadows    1
                       ..
1054 Zappia Drive       1
2043 Jadewood Drive     1
3539 Bottom Lane        1
2886 Straford Park      1
570 Alpha Avenue        1
Name: address, Length: 483, dtype: int64

In [233]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [234]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [235]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

#### 2-treatments table

In [236]:
treatments.sample(3)

given_name surname    auralin novodra  hba1c_start  hba1c_end  \
81        robert  wagner  43u - 49u       -         9.84       9.52   
212    alexander  hueber  35u - 38u       -         7.71       7.35   
22   brancaleone   russo  53u - 60u       -         8.61       8.18   

     hba1c_change  
81           0.32  
212          0.36  
22            NaN

In [237]:
treatments.shape

(280, 7)

In [238]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 11.0+ KB


In [239]:
treatments[treatments.duplicated()]

given_name surname    auralin novodra  hba1c_start  hba1c_end  \
136     joseph     day  29u - 36u       -          7.7       7.19   

     hba1c_change  
136           NaN

In [240]:
treatments.dtypes

given_name       object
surname          object
auralin          object
novodra          object
hba1c_start     float64
hba1c_end       float64
hba1c_change    float64
dtype: object

In [241]:
var=list(treatments.columns)
for item in var:
    null_value_incol = treatments[item].isnull().sum()
    print(item+":"+str(null_value_incol))

given_name:0
surname:0
auralin:0
novodra:0
hba1c_start:0
hba1c_end:0
hba1c_change:109


In [242]:
treatments.sample(10)

given_name      surname    auralin    novodra  hba1c_start  \
239             linda        lundy          -  48u - 53u         7.97   
259           aaliyah         rice          -  31u - 31u         7.64   
88             yumena     nakayama          -  34u - 32u         7.76   
220            sinead       reilly  27u - 39u          -         7.76   
176             andré       nordin  40u - 47u          -         7.60   
276              john  teichelmann          -  49u - 49u         7.90   
82               maya     isaksson  33u - 41u          -         7.66   
120  zikoranaudodimma     chinedum  41u - 51u          -         8.19   
136            joseph          day  29u - 36u          -         7.70   
107          flaviano     endrizzi          -  46u - 43u         7.93   

     hba1c_end  hba1c_change  
239       7.52          0.95  
259       7.33          0.31  
88        7.35          0.91  
220       7.39          0.37  
176       7.27          0.33  
276       7.58           NaN  
82        7.17           NaN  
120       7.83          0.36  
136       7.19           NaN  
107       7.56          0.37

In [243]:
var=list(treatments.columns)
for item in var:
    duplicate=treatments[item].duplicated().sum()
    print(item +":"+str(duplicate))

given_name:7
surname:8
auralin:159
novodra:167
hba1c_start:183
hba1c_end:171
hba1c_change:249


In [244]:
treatments[treatments.duplicated()]

given_name surname    auralin novodra  hba1c_start  hba1c_end  \
136     joseph     day  29u - 36u       -          7.7       7.19   

     hba1c_change  
136           NaN

In [245]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

#### 3-adverse reactions table

In [246]:
adverse_reactions.sample(3)

given_name     surname           adverse_reaction
16      jakob    jakobsen               hypoglycemia
21   miłosław  wiśniewski  injection site discomfort
15     jia li        teng                     nausea

In [247]:
adverse_reactions.dtypes

given_name          object
surname             object
adverse_reaction    object
dtype: object

In [248]:
adverse_reactions.shape

(34, 3)

In [249]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 472.0+ bytes


In [250]:
var=list(adverse_reactions.columns)
for item in var:
    null_value_incol = adverse_reactions[item].isnull().sum()
    print(item+":"+str(null_value_incol))

given_name:0
surname:0
adverse_reaction:0


In [251]:
adverse_reactions.sample(34)

given_name     surname           adverse_reaction
1          lena        baer               hypoglycemia
20     anenechi       chidi               hypoglycemia
14     gabriele     saenger               hypoglycemia
10        tegan     johnson                   headache
29         alex    crawford               hypoglycemia
2        joseph         day               hypoglycemia
15       jia li        teng                     nausea
21     miłosław  wiśniewski  injection site discomfort
31       steven         roy                   headache
13         leon      scholz  injection site discomfort
4       manouck     wubbels          throat irritation
32      cecilie      nilsen               hypoglycemia
3        flavia  fiorentino                      cough
7       albinca     komavec               hypoglycemia
0         berta  napolitani  injection site discomfort
28       xiuxiu       chang               hypoglycemia
27       idalia       moore               hypoglycemia
8           noe      aranda               hypoglycemia
12    abdul-nur         isa               hypoglycemia
18          ole    petersen               hypoglycemia
9         sofia   hermansen  injection site discomfort
33    krisztina      magyar               hypoglycemia
22        lixue       hsueh  injection site discomfort
17  christopher    woodward                     nausea
26      clinton      miller          throat irritation
24         kang         mai  injection site discomfort
16        jakob    jakobsen               hypoglycemia
25       elliot  richardson               hypoglycemia
11         abel     yonatan                      cough
5       jasmine       sykes               hypoglycemia
23        merci      leroux               hypoglycemia
19       finley    chandler                   headache
6        louise     johnson               hypoglycemia
30       monika      lončar               hypoglycemia

In [252]:
var=list(adverse_reactions.columns)
for item in var:
    duplicate=adverse_reactions[item].duplicated().sum()
    print(item +":"+str(duplicate))

given_name:0
surname:1
adverse_reaction:28


In [253]:
adverse_reactions.duplicated().sum()

0

In [254]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]


14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [255]:
adverse_reactions.describe()

given_name  surname adverse_reaction
count          34       34               34
unique         34       33                6
top        monika  johnson     hypoglycemia
freq            1        2               19

### -----------------------------------------------------------------------------------------------------------------------
# Documenting quality and tideness issues

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [256]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

#### `treatments`: Missing records (280 instead of 350)

##### Code

In [257]:
treatments_clean_missing=pd.DataFrame(pd.read_csv('treatments_cut.csv'))
treatments_clean=pd.concat([treatments_clean_missing,treatments_clean],ignore_index=True)


##### Test

In [258]:
treatments_clean

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0        jožka     resanovič  22u - 30u          -         7.56       7.22   
1    inunnguaq      heilmann  57u - 67u          -         7.85       7.45   
2        alwin      svensson  36u - 39u          -         7.78       7.34   
3          thể         lương          -  61u - 64u         7.64       7.22   
4       amanda       ribeiro  36u - 44u          -         7.85       7.47   
..         ...           ...        ...        ...          ...        ...   
345     albina      zetticci  45u - 51u          -         7.93       7.73   
346       john   teichelmann          -  49u - 49u         7.90       7.58   
347     mathea       lillebø  23u - 36u          -         9.04       8.67   
348     vallie        prince  31u - 38u          -         7.64       7.28   
349     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0            0.34  
1             NaN  
2             NaN  
3            0.92  
4            0.38  
..            ...  
345          0.20  
346           NaN  
347          0.37  
348          0.36  
349          0.36  

[350 rows x 7 columns]

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)(accuracy issue)


##### Define 
comlete those nan values by filling it with subtraction of HbA1c_start and HbA1c_end

##### Code

In [259]:
treatments_clean['hba1c_change']=treatments_clean['hba1c_start']-treatments_clean['hba1c_end']


##### Test

In [260]:
treatments_clean['hba1c_change'].isnull().sum()

0

### Tidiness

#### Contact column in `patients` table contains two variables: phone number and email

##### Code

In [261]:
patients_clean['email']=patients_clean['contact'].str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])',expand=True)
patients_clean.sample(10)

patient_id assigned_sex given_name  surname                   address  \
102         103       female      Najla    Touma  2146 Willow Greene Drive   
186         187       female       Jane  Citizen          2866 Myra Street   
244         245         male       John      Doe           123 Main Street   
286         287         male      Lewis     Webb                       NaN   
73           74       female      Žarka      Rap       2661 Russell Street   
294         295       female      Annie    Allen          3634 Lyon Avenue   
158         159         male    Chibuzo    Okoli          4291 Patton Lane   
246         247       female       Meya  Sjöberg       1233 Liberty Avenue   
250         251       female        Mee    Chung     3084 Blue Spruce Lane   
441         442         male       Luis    Silva         959 Heritage Road   

            city       state  zip_code        country  \
102   Montgomery          AL   36104.0  United States   
186   Providence          RI    2908.0  United States   
244     New York          NY   12345.0  United States   
286          NaN         NaN       NaN            NaN   
73       Foxboro          MA    2035.0  United States   
294    Cambridge          MA    2142.0  United States   
158        Tulsa          OK   74136.0  United States   
246    Santa Ana  California   92704.0  United States   
250      Dundalk          MD   21222.0  United States   
441  San Joaquin          CA   93660.0  United States   

                                         contact   birthdate  weight  height  \
102   NajlaQismahTouma@cuvox.de+1 (334) 261-1235   8/20/1948   165.0      61   
186        JaneCitizen@cuvox.de+1 (401) 485-6384   9/27/1974   143.0      65   
244                  johndoe@email.com1234567890    1/1/1975   180.0      72   
286                                          NaN    4/1/1979   155.3      68   
73            978-969-6753ZarkaRap@superrito.com  12/28/1995   139.3      68   
294      AnnieJAllen@superrito.com1 508 921 6327   3/31/1926   159.7      60   
158     ChibuzoOkoli@einrot.com+1 (918) 971-5864   6/14/1983   142.8      73   
246             714-431-2746MeyaSjoberg@cuvox.de   5/15/1968   232.1      66   
250             410-284-8935MeeChung@teleworm.us  10/27/1932   206.1      64   
441  559-693-6779LuisRibeiroSilva@jourrapide.com   11/2/1946   239.1      71   

      bmi                            email  
102  31.2        NajlaQismahTouma@cuvox.de  
186  23.8             JaneCitizen@cuvox.de  
244  24.4                johndoe@email.com  
286  23.6                              NaN  
73   21.2           ZarkaRap@superrito.com  
294  31.2        AnnieJAllen@superrito.com  
158  18.8          ChibuzoOkoli@einrot.com  
246  37.5             MeyaSjoberg@cuvox.de  
250  35.4             MeeChung@teleworm.us  
441  33.3  LuisRibeiroSilva@jourrapide.com

##### Test

In [262]:
patients_clean['phonenumber']=patients_clean['contact'].str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})',expand=True)
patients_clean.sample(9)


patient_id assigned_sex given_name     surname  \
283         284         male  Nwachukwu   Nebeolisa   
90           91         male       Ingo      Rokavc   
109         110         male    Stephen    Mayberry   
389         390         male      Yegor    Uspensky   
387         388       female     Vallie      Prince   
275         276         male      Eddie      Archer   
241         242       female       Skye  Gormanston   
23           24         male      Lovre       Galić   
439         440         male     Hiromu    Horikawa   

                       address             city     state  zip_code  \
283     2873 John Calvin Drive          Chicago        IL   60605.0   
90     4271 Cherry Ridge Drive          Buffalo  New York   14214.0   
109     3063 School House Road      Hattiesburg        MS   39402.0   
389           4900 Philli Lane            Tulsa        OK   74133.0   
387          550 Mulberry Lane  Fort Lauderdale        FL   33301.0   
275        2043 Jadewood Drive          Lombard  Illinois   60148.0   
241  936 Lightning Point Drive        Arlington        TN   38002.0   
23           4941 Marion Drive     Winter Haven   Florida   33830.0   
439      4567 Hazelwood Avenue       Des Moines        IA   50309.0   

           country                                       contact  birthdate  \
283  United States  NwachukwuNebeolisa@cuvox.de+1 (708) 845-2053  3/10/1986   
90   United States     IngoRokavc@superrito.com+1 (585) 902-9127  8/27/1977   
109  United States   601-699-4153StephenFMayberry@jourrapide.com   9/1/1934   
389  United States         918-459-9811YegorUspensky@fleckens.hu  9/10/1956   
387  United States            561-826-5683VallieSPrince@cuvox.de  1/16/1998   
275  United States       EddieAArcher@gustr.com+1 (224) 305-6805  7/17/1982   
241  United States         901-484-5225SkyeGormanston@dayrep.com  1/20/1934   
23   United States            LovreGalic@gustr.com1 813 355 9476  5/26/1960   
439  United States    HiromuHorikawa@superrito.com1 515 362 6574   7/7/1981   

     weight  height   bmi                            email        phonenumber  
283   245.5      68  37.3      NwachukwuNebeolisa@cuvox.de  +1 (708) 845-2053  
90    192.3      67  30.1         IngoRokavc@superrito.com  +1 (585) 902-9127  
109   166.1      72  22.5  StephenFMayberry@jourrapide.com       601-699-4153  
389   136.8      65  22.8        YegorUspensky@fleckens.hu       918-459-9811  
387   152.7      67  23.9           VallieSPrince@cuvox.de       561-826-5683  
275   158.6      69  23.4           EddieAArcher@gustr.com  +1 (224) 305-6805  
241   139.7      60  27.3        SkyeGormanston@dayrep.com       901-484-5225  
23    222.9      66  36.0             LovreGalic@gustr.com       813 355 9476  
439   242.0      70  34.7     HiromuHorikawa@superrito.com       515 362 6574

In [263]:
patients_clean[patients_clean['phonenumber'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi email  \
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2   NaN   
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1   NaN   
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4   NaN   
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2   NaN   
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4   NaN   
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6   NaN   
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2   NaN   
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1   NaN   
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1   NaN   
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4   NaN   
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6   NaN   
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1   NaN   

    phonenumber  
209         NaN  
219         NaN  
230         NaN  
234         NaN  
242         NaN  
249         NaN  
257         NaN  
264         NaN  
269         NaN  
278         NaN  
286         NaN  
296         NaN

In [264]:
patients_clean.drop('contact', axis='columns', inplace=True)

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Code

In [265]:
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name','surname','hba1c_start','hba1c_end','hba1c_change'], 
                 var_name="treatment", value_name="total_dose")
treatments_clean=treatments_clean[treatments_clean['total_dose']!='-']
treatments_clean['start_dose'] , treatments_clean['end_dose']=treatments_clean['total_dose'].str.split(' - ', 1).str
treatments_clean

C:\Users\ENTER\anaconda3\lib\site-packages\ipykernel_launcher.py:4: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  after removing the cwd from sys.path.


given_name      surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0        jožka    resanovič         7.56       7.22          0.34   auralin   
1    inunnguaq     heilmann         7.85       7.45          0.40   auralin   
2        alwin     svensson         7.78       7.34          0.44   auralin   
4       amanda      ribeiro         7.85       7.47          0.38   auralin   
6         sara        miles         7.53       7.12          0.41   auralin   
..         ...          ...          ...        ...           ...       ...   
688     yasmin       araujo         7.82       7.36          0.46   novodra   
689     hiromu     horikawa         7.77       7.28          0.49   novodra   
691        leo       vieira         7.74       7.36          0.38   novodra   
692     steven          roy         7.87       7.43          0.44   novodra   
696       john  teichelmann         7.90       7.58          0.32   novodra   

    total_dose start_dose end_dose  
0    22u - 30u        22u      30u  
1    57u - 67u        57u      67u  
2    36u - 39u        36u      39u  
4    36u - 44u        36u      44u  
6    30u - 35u        30u      35u  
..         ...        ...      ...  
688  51u - 54u        51u      54u  
689  47u - 46u        47u      46u  
691  30u - 33u        30u      33u  
692  41u - 43u        41u      43u  
696  49u - 49u        49u      49u  

[350 rows x 9 columns]

##### Test

In [266]:
treatments_clean=treatments_clean.drop('total_dose',axis=1)
treatments_clean


given_name      surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0        jožka    resanovič         7.56       7.22          0.34   auralin   
1    inunnguaq     heilmann         7.85       7.45          0.40   auralin   
2        alwin     svensson         7.78       7.34          0.44   auralin   
4       amanda      ribeiro         7.85       7.47          0.38   auralin   
6         sara        miles         7.53       7.12          0.41   auralin   
..         ...          ...          ...        ...           ...       ...   
688     yasmin       araujo         7.82       7.36          0.46   novodra   
689     hiromu     horikawa         7.77       7.28          0.49   novodra   
691        leo       vieira         7.74       7.36          0.38   novodra   
692     steven          roy         7.87       7.43          0.44   novodra   
696       john  teichelmann         7.90       7.58          0.32   novodra   

    start_dose end_dose  
0          22u      30u  
1          57u      67u  
2          36u      39u  
4          36u      44u  
6          30u      35u  
..         ...      ...  
688        51u      54u  
689        47u      46u  
691        30u      33u  
692        41u      43u  
696        49u      49u  

[350 rows x 8 columns]

#### Adverse reaction should be part of the `treatments` table

##### Code

In [267]:
treatments_clean=pd.merge(treatments_clean,adverse_reactions_clean,on=['given_name','surname'],how='left')
treatments_clean

given_name      surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0        jožka    resanovič         7.56       7.22          0.34   auralin   
1    inunnguaq     heilmann         7.85       7.45          0.40   auralin   
2        alwin     svensson         7.78       7.34          0.44   auralin   
3       amanda      ribeiro         7.85       7.47          0.38   auralin   
4         sara        miles         7.53       7.12          0.41   auralin   
..         ...          ...          ...        ...           ...       ...   
345     yasmin       araujo         7.82       7.36          0.46   novodra   
346     hiromu     horikawa         7.77       7.28          0.49   novodra   
347        leo       vieira         7.74       7.36          0.38   novodra   
348     steven          roy         7.87       7.43          0.44   novodra   
349       john  teichelmann         7.90       7.58          0.32   novodra   

    start_dose end_dose adverse_reaction  
0          22u      30u              NaN  
1          57u      67u              NaN  
2          36u      39u              NaN  
3          36u      44u              NaN  
4          30u      35u              NaN  
..         ...      ...              ...  
345        51u      54u              NaN  
346        47u      46u              NaN  
347        30u      33u              NaN  
348        41u      43u         headache  
349        49u      49u              NaN  

[350 rows x 9 columns]

##### Test

In [268]:
treatments_clean.shape

(350, 9)

In [269]:
id_names = patients_clean[['given_name', 'surname','patient_id']]
id_names.given_name = id_names.given_name.str.lower()
id_names.surname = id_names.surname.str.lower()
treatments_clean = pd.merge(treatments_clean, id_names, on=['given_name', 'surname'])
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)
treatments_clean

C:\Users\ENTER\anaconda3\lib\site-packages\pandas\core\generic.py:5303: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


hba1c_start  hba1c_end  hba1c_change treatment start_dose end_dose  \
0           7.56       7.22          0.34   auralin        22u      30u   
1           7.85       7.45          0.40   auralin        57u      67u   
2           7.78       7.34          0.44   auralin        36u      39u   
3           7.85       7.47          0.38   auralin        36u      44u   
4           7.53       7.12          0.41   auralin        30u      35u   
..           ...        ...           ...       ...        ...      ...   
344         7.82       7.36          0.46   novodra        51u      54u   
345         7.77       7.28          0.49   novodra        47u      46u   
346         7.74       7.36          0.38   novodra        30u      33u   
347         7.87       7.43          0.44   novodra        41u      43u   
348         7.90       7.58          0.32   novodra        49u      49u   

    adverse_reaction  patient_id  
0                NaN         468  
1                NaN         107  
2                NaN         438  
3                NaN         144  
4                NaN         260  
..               ...         ...  
344              NaN          59  
345              NaN         440  
346              NaN         434  
347         headache         394  
348              NaN         141  

[349 rows x 8 columns]

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Code

##### Test

In [270]:
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

22    patient_id
dtype: object

### Quality

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Code

In [271]:
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5, fillchar='0')
# Reconvert NaNs entries that were converted to '0000n' by code above
patients_clean.zip_code = patients_clean.zip_code.replace('0000n', np.nan)


##### Test

In [272]:
patients_clean.zip_code.head()

0    92390
1    61812
2    68467
3    07095
4    36303
Name: zip_code, dtype: object

#### Tim Neudorf height is 27 in instead of 72 in

##### Code

In [273]:
patients_clean.height=patients_clean.height.replace(27,72)

##### Test

In [274]:
patients_clean.query('height==27')

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, email, phonenumber]
Index: []

#### Full state names sometimes, abbreviations other times

##### Code

In [275]:
state_abbrev = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}

# Function to apply
def abbreviate_state(patient):
    if patient['state'] in state_abbrev.keys():
        abbrev = state_abbrev[patient['state']]
        return abbrev
    else:
        return patient['state']
    
patients_clean['state'] = patients_clean.apply(abbreviate_state, axis=1)


##### Test

In [276]:
patients_clean['state'].value_counts()

CA    60
NY    47
TX    32
IL    24
MA    22
FL    22
PA    18
GA    15
OH    14
OK    13
MI    13
LA    13
NJ    12
VA    11
MS    10
WI    10
TN     9
AL     9
IN     9
MN     9
NC     8
WA     8
KY     8
MO     7
KS     6
ID     6
NV     6
NE     6
CT     5
SC     5
IA     5
CO     4
ND     4
AZ     4
ME     4
RI     4
AR     4
SD     3
WV     3
DE     3
MD     3
OR     3
MT     2
VT     2
DC     2
NM     1
NH     1
AK     1
WY     1
Name: state, dtype: int64

#### Dsvid Gustafsson

##### Define
*wrong spelling David.*

##### Code

In [277]:
patients_clean.given_name=patients_clean.given_name.replace('Dsvid','David')

##### Test

In [278]:
patients_clean[patients_clean.given_name == 'David']

patient_id assigned_sex given_name     surname               address  \
8             9         male      David  Gustafsson    1790 Nutter Street   
280         281         male      David    Beauvais    98 Freshour Circle   
440         441         male      David   Villadsen  1434 Oak Ridge Drive   

             city state zip_code        country   birthdate  weight  height  \
8     Kansas City    MO    64105  United States    3/6/1937   163.9      66   
280   San Antonio    TX    78258  United States    5/1/1931   128.5      67   
440  St Elizabeth    MO    65559  United States  11/30/1924   137.3      68   

      bmi                        email        phonenumber  
8    26.5  DavidGustafsson@armyspy.com       816-265-9578  
280  20.1     DavidLBeauvais@rhyta.com       210-218-3477  
440  20.9  DavidMVilladsen@armyspy.com  +1 (573) 493-4748

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Code

In [279]:
# To category
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')

# To datetime
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

# Strip u and to integer
treatments_clean.start_dose = treatments_clean.start_dose.str.strip('u').astype(int)
treatments_clean.end_dose = treatments_clean.end_dose.str.strip('u').astype(int)

In [280]:
treatments_clean.sample(2)

hba1c_start  hba1c_end  hba1c_change treatment  start_dose  end_dose  \
25          7.58       7.29          0.29   auralin          43        56   
102         7.79       7.49          0.30   auralin          54        54   

    adverse_reaction  patient_id  
25               NaN         395  
102              NaN         374

#### Multiple phone number formats

##### Code

In [281]:
# Your cleaning code here
patients_clean.phonenumber = patients_clean.phonenumber.str.replace(r'\D+', '').str.pad(11, fillchar='1')

##### Test

In [282]:
# Your testing code here
patients_clean.phonenumber

0      19517199170
1      12175693204
2      14023636804
3      17326368246
4      13345157487
          ...     
498    12074770579
499    19282844492
500    18162236007
501    13604432060
502    14028484923
Name: phonenumber, Length: 503, dtype: object

#### Default John Doe data

##### Define
*Recall that it is assumed that the data that this John Doe data displaced is not recoverable,so removing it ,is better.*
so removing it ,is better

##### Code

In [283]:
patients_clean = patients_clean[patients_clean.surname != 'Doe']

##### Test

In [284]:
patients_clean.surname.value_counts()

Jakobsen       3
Taylor         3
Tạ             2
Kadyrov        2
Nilsen         2
              ..
Rukavina       1
Konovšek       1
Kozłowska      1
Villalpando    1
Vukelić        1
Name: surname, Length: 465, dtype: int64

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
*removing any duplicate data.*

##### Code

In [285]:
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

##### Test

In [286]:
patients_clean[patients_clean.surname == 'Jakobsen']

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
432      Houston    TX    77020  United States 1962-11-25   185.2      67   

      bmi                         email  phonenumber  
24   24.4     JakobCJakobsen@einrot.com  18458587707  
432  29.0  KarenJakobsen@jourrapide.com  19792030438

In [287]:
patients_clean[patients_clean.surname == 'Taylor']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

    state zip_code        country  birthdate  weight  height   bmi  \
131    WV    25962  United States 1960-10-23   206.1      64  35.4   
426    FL    33179  United States 1992-09-02   186.6      69  27.6   

                          email  phonenumber  
131    SandraCTaylor@dayrep.com  13044382648  
426  RogelioJTaylor@teleworm.us  13054346299

#### kgs instead of lbs for Zaitseva weight

##### Define
*converting weight from kg to ibs for **Zaitseva** as his weight is 48.7.*

##### Code

In [288]:
min_weight=patients_clean.weight.min()
patients_clean[patients_clean.weight==min_weight]
patients_clean.weight=patients_clean.weight.replace(48.8,(48.8*2.20462))


##### Test

In [289]:
patients_clean[patients_clean.weight==min_weight]


Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, email, phonenumber]
Index: []

## saving all cleaned files to be ready for analysis

In [290]:
patients_clean.to_csv('patients_clean.csv',index=False)
treatments_clean.to_csv('treatments_clean.csv',index=False)

**END CLEANING**